File: bombe/submissions/accepted/Alexis.cpp
#include <iostream>
#include <iomanip>
#include <cmath>
#include <unordered_map>
#include <set>
#include <vector>

#define int long long

using namespace std;


//TODO: CHECK IF BAD CONSTANT TO BE FAIR DURING THE CONTEST
int MOD = (1LL << 40);
int A = 911382323;

int safeMod(int a, int m) {
    return (a % m + m) % m;
}

void solve() {

    int n, m; cin >> n >> m;
    vector<vector<char>> a(n, vector<char>(m));
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
            cin >> a[i][j];
        }
    }

    vector<int> pre(m);

    pre[0] = 1;
    for(int i = 1; i < m; i++) {
        pre[i] = safeMod((pre[i-1] * A), MOD);
    }
    vector<int> hs(n);
    unordered_map<int, pair<int, int>> cnt;

    for(int i = 0; i < n; i++) {
        int hash = 0;

        for(int j = 0; j < m; j++) {
            hash += a[i][j] * pre[j];
            hash = safeMod(hash, MOD);
        }
        hs[i] = hash;

        cnt[hs[i]].first = i+1;
    }

    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
            hs[i] -= (a[i][m-j-1] * pre[m-1])% MOD;
            hs[i] %= MOD;
            hs[i] *= A;
            hs[i] %= MOD;
            hs[i] += a[i][m-j-1];
            hs[i] %= MOD;

            if(cnt[hs[i]].first == 0){
                cnt[hs[i]].first = i+1;
            } else if (cnt[hs[i]].second == 0 && cnt[hs[i]].first != (i+1)) {
                cnt[hs[i]].second = i+1;
            } else if(cnt[hs[i]].first > i+1) {
                cnt[hs[i]].first = i+1;
            } else if(cnt[hs[i]].second > i+1) {
                cnt[hs[i]].second = i+1;
            }
        }
    }

    pair<int, int> mini = make_pair(10e9, 10e9);

    for(auto& k : cnt) {
        if(k.second.first == 0 || k.second.second == 0) continue;
        if(k.second < mini) mini = k.second;
    }

    if(mini.first == 10e9 && mini.second == 10e9) {
        cout << "BOOM!\n";
    } else {
        cout << mini.first << " " << mini.second << endl;
    }
}

signed main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    solve();

    return 0;
}

File: bombe/submissions/accepted/alexis.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;
import java.util.StringTokenizer;

public class alexis {

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                if (st.hasMoreTokens()) {
                    str = st.nextToken("\n");
                } else {
                    str = br.readLine();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }

    static final long MOD = (1L << 40);
    static final long A = 911382323;

    static long safeMod(long a, long m) {
        return (a % m + m) % m;
    }

    static void solve() {
        FastReader scanner = new FastReader();

        int n = scanner.nextInt();
        int m = scanner.nextInt();

        char[][] a = new char[n][m];

        for (int i = 0; i < n; i++) {
            String row = scanner.next();
            for (int j = 0; j < m; j++) {
                a[i][j] = row.charAt(j);
            }
        }

        long[] pre = new long[m];
        pre[0] = 1;

        for (int i = 1; i < m; i++) {
            pre[i] = safeMod(pre[i - 1] * A, MOD);
        }

        long[] hs = new long[n];
        Map<Long, Pair<Long, Long>> cnt = new HashMap<>();

        for (int i = 0; i < n; i++) {
            long hash = 0;

            for (int j = 0; j < m; j++) {
                hash += a[i][j] * pre[j];
                hash = safeMod(hash, MOD);
            }
            hs[i] = hash;

            cnt.put(hs[i], new Pair<>(i + 1L, 0L));
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                hs[i] -= (a[i][m - j - 1] * pre[m - 1]) % MOD;
                hs[i] %= MOD;
                hs[i] *= A;
                hs[i] %= MOD;
                hs[i] += a[i][m - j - 1];
                hs[i] %= MOD;

                if(cnt.get(hs[i]) == null) {
                    cnt.put(hs[i], new Pair<>(0L, 0L));
                }
                Pair<Long, Long> pair = cnt.get(hs[i]);

                if (pair.first == 0L) {
                    pair.first = i + 1L;
                } else if (pair.second == 0L && pair.first != (i + 1L)) {
                    pair.second = i + 1L;
                } else if (pair.first > i + 1L) {
                    pair.first = i + 1L;
                } else if (pair.second > i + 1L) {
                    pair.second = i + 1L;
                }

                cnt.replace(hs[i], pair);

            }
        }

        Pair<Long, Long> mini = new Pair<>(10_000_000_000L, 10_000_000_000L);

        for (Map.Entry<Long, Pair<Long, Long>> entry : cnt.entrySet()) {
            Pair<Long, Long> value = entry.getValue();

            if (value.first == 0 || value.second == 0) continue;
            if (value.compareTo(mini) < 0) mini = value;
        }

        if (mini.first == 10_000_000_000L && mini.second == 10_000_000_000L) {
            System.out.println("BOOM!");
        } else {
            System.out.println(mini.first + " " + mini.second);
        }
    }

    public static void main(String[] args) {
        solve();
    }

    static class Pair<T extends Comparable<T>, U extends Comparable<U>> implements Comparable<Pair<T, U>> {
        T first;
        U second;

        Pair(T first, U second) {
            this.first = first;
            this.second = second;
        }

        @Override
        public int compareTo(Pair<T, U> other) {
            int cmp = this.first.compareTo(other.first);
            if (cmp != 0) {
                return cmp;
            }
            return this.second.compareTo(other.second);
        }
    }
}
